helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java

```diff
@@ -814,6 +837,93 @@ private static AvroVersion detectAvroVersion() {
+    boolean fileResourceLoaderClassExists = checkIfVelocityEngineFileResourceLoaderClassExists(specificCompilerClass);
+
+    // SpecificCompiler.hasBuilder(Schema s) method is added in 1.8.0
+    try {
+      specificCompilerClass.getMethod("hasBuilder", Schema.class);
+    } catch (NoSuchMethodException expected) {
+      if (fileResourceLoaderClassExists) {
+        return AvroVersion.AVRO_1_7;
+      } else {
+        return AvroVersion.AVRO_1_6;
+      }
+    }
+
+    //SpecificCompiler.isGettersReturnOptional() method added in 1.9.0
+    try {
+      specificCompilerClass.getMethod("isGettersReturnOptional");
+    } catch (NoSuchMethodException expected) {
+      return AvroVersion.AVRO_1_8;
+    }
+
+    //SpecificCompiler.isOptionalGettersForNullableFieldsOnly() method added in 1.10.0
+    try {
+      specificCompilerClass.getMethod("isOptionalGettersForNullableFieldsOnly");
+    } catch (NoSuchMethodException expected) {
+      return AvroVersion.AVRO_1_9;
+    }
+
+    // SpecificCompiler.getUsedCustomLogicalTypeFactories() method added in 1.11.0
+    try {
+      specificCompilerClass.getMethod("getUsedCustomLogicalTypeFactories", Schema.class);
+    } catch (NoSuchMethodException expected) {
+      return AvroVersion.AVRO_1_10;
+    }
+
+    return AvroVersion.AVRO_1_11;
+  }
+
+  // We are fishing for the ```file.resource.loader.class``` property in the private field `velocityEngine` of
+  // a SpecificCompiler class.
+  private static boolean checkIfVelocityEngineFileResourceLoaderClassExists(Class<?> specificCompilerClass) {
+    try {
+      // If `velocityEngine` were public, we could simplify this reflection to be
+      // ```VelocityEngine velocityEngine = (new SpecificRecord()).velocityEngine;````
+      Field velocityEngineField = specificCompilerClass.getDeclaredField("velocityEngine");
+      velocityEngineField.setAccessible(true);
+      // The default constructor is also private.
+      Constructor constructor = specificCompilerClass.getDeclaredConstructor();
+      constructor.setAccessible(true);
+      constructor.newInstance();
+      Object velocityEngine = velocityEngineField.get(constructor.newInstance());
+
+      // Without reflection, this would be ```velocityEngine.getProperty("file.resource.loader.class");```
+      Class<?> velocityEngineClass = Class.forName("org.apache.velocity.app.VelocityEngine");
+      Method velocityEngineGetPropertyMethod = velocityEngineClass.getMethod("getProperty", String.class);
+      Object fileResourceLoaderClassProperty = velocityEngineGetPropertyMethod.invoke(velocityEngine, "file.resource.loader.class");
+
+      return fileResourceLoaderClassProperty != null;
+
+      // We should never reach these exceptions
+    } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | ClassNotFoundException | InvocationTargetException | NoSuchFieldException expected) {
+      throw new RuntimeException(expected.fillInStackTrace());
```

Comment:
lmk if there's a better exception to raise here...

Reply:
i recommend throwing something like IllegalStateException with the msg "unexpected exception while retrieving velocity engine config" and supplying the caught exception as a cause.

Reply:
you should also be able to simplify this to catch(Exception) {} i think ?